src/Compilers/CSharp/Portable/Binder/UsingStatementBinder.cs

```diff
@@ -149,6 +121,69 @@ internal override BoundStatement BindUsingStatementParts(DiagnosticBag diagnosti
                 boundBody,
                 awaitOpt,
                 hasErrors);
+
+            // returns true for error
+            bool initConversionConsideringAlternate(TypeSymbol disposableInterface, DiagnosticBag bag, bool fromExpression)
+            {
+                DisposableConversion conversionResult = getConversion(disposableInterface, bag, fromExpression, out iDisposableConversion);
+
+                switch (conversionResult)
+                {
+                    case DisposableConversion.CascadingError:
+                        return true;
+                    case DisposableConversion.FailedNotReported:
+                        // Retry with a different assumption about whether the `using` is async
+                        TypeSymbol alternateInterface = getDisposableInterface(!hasAwait);
+                        var alternateConversionResult = getConversion(alternateInterface, bag: null, fromExpression: expressionSyntax != null, out _);
+
+                        bool wrongAsync = (alternateConversionResult == DisposableConversion.Succeeded);
+                        ErrorCode errorCode = wrongAsync
+                            ? (hasAwait ? ErrorCode.ERR_NoConvToIAsyncDispWrongAsync : ErrorCode.ERR_NoConvToIDispWrongAsync)
+                            : (hasAwait ? ErrorCode.ERR_NoConvToIAsyncDisp : ErrorCode.ERR_NoConvToIDisp);
+
+                        Error(diagnostics, errorCode, (CSharpSyntaxNode)declarationSyntax ?? expressionSyntax, declarationTypeOpt ?? expressionOpt.Display);
+                        return true;
+                    default:
+                        return false;
+                }
+            }
+
+            DisposableConversion getConversion(TypeSymbol disposableInterface, DiagnosticBag bag, bool fromExpression, out Conversion conversion)
```

Comment:
>DisposableConversion getConversion(TypeSymbol disposableInterface, DiagnosticBag bag, bool fromExpression, out Conversion conversion) [](start = 12, length = 133)

It feels that the refactoring in combination with the need for a new enum, makes things more complex than they could be. On the second use we only need to check if ```conversion.IsImplicit``` and that second use can be inlined into the if that right now returns DisposableConversion.FailedNotReported and that if can be moved to initConversionConsideringAlternate.  Basically, we only need the helper that does:
```
                conversion = fromExpression ?
                    originalBinder.Conversions.ClassifyImplicitConversionFromExpression(expressionOpt, disposableInterface, ref useSiteDiagnostics) :
                    originalBinder.Conversions.ClassifyImplicitConversionFromType(declarationTypeOpt, disposableInterface, ref useSiteDiagnostics);
```

Reply:
That was simpler indeed. Thanks!